packages/block-events-stream/src/createBlockNumberStream.ts

```diff
@@ -0,0 +1,44 @@
+import { BehaviorSubject, filter, map } from "rxjs";
+import type { Block, BlockNumber, BlockTag, PublicClient } from "viem";
+import { createBlockStream } from "./createBlockStream";
+import { ReadonlyBehaviorSubject } from "./common";
+import { isNonPendingBlock } from "./isNonPendingBlock";
+
+// TODO: pass through viem's types, e.g. WatchBlocksParameters -> GetBlockReturnType
+
+export type CreateBlockNumberStreamOptions =
+  | {
+      publicClient: PublicClient;
+      blockTag: Omit<BlockTag, "pending">;
+      block$?: never;
+    }
+  | {
+      publicClient?: never;
+      blockTag?: never;
+      block$: ReadonlyBehaviorSubject<Block>;
+    };
+
+export async function createBlockNumberStream({
+  publicClient,
+  blockTag,
+  block$: initialBlock$,
+}: CreateBlockNumberStreamOptions): Promise<ReadonlyBehaviorSubject<BlockNumber>> {
+  const block$ = initialBlock$ ?? (await createBlockStream({ publicClient, blockTag: blockTag as BlockTag }));
+  const block = block$.value;
+
+  if (!block.number) {
+    // TODO: better error
+    throw new Error(`${blockTag} block missing or pending`);
```

Comment:
what does it mean if a block we receive here is pending? Is this error just for if the initial block of our stream is not available?

Reply:
yep! it's more edge-casey and to help TS infer the proper types and bypass the "block number can be null for pending blocks"